txnkv/transaction/2pc.go

```diff
@@ -1349,7 +1349,7 @@ func keepAlive(
 			}
 
 			// broadcast to all stores
-			if isPipelinedTxn {
+			if err != nil && isPipelinedTxn {
```

Comment:
Isn't it `if err == nil ...`?

Maybe add a comment to explain why skip txn status broadcast when heartbeat failed.

Maybe this way is better:

```go
if err != nil {
	... // logs
	if stopHeartBeat || keepFail > maxConsecutiveFailure {
		return
	}
	continue
}

// heartbeat success
keepFail = 0
metrics.TxnHeartBeatHistogramOK.Observe(time.Since(startTime).Seconds())
if isPipelinedTxn {
	... // broadcast
}